src/test/resources/TestNG.xml

```diff
@@ -7,7 +7,9 @@
         <parameter name="BaseURL" value="https://bbb.testpro.io/" />
         <classes>
             <class name="LoginTestRefactored"/>
-            <class name="MainPageTestRefactored"/>
+<!--            <class name="MainPageTestRefactored"/>-->
+<!--            <class name="LoginTests"/>-->
+<!--            <class name="Homework20"/>-->
         </classes>
     </test>
 </suite>
```

Comment:
The code is not compiled! Please finish the homework!You got this problem, please correct it. 

![image](https://user-images.githubusercontent.com/104761482/221134397-b514dd55-b9dc-4d66-8428-6d1a32d322c4.png)

**HINT**

**Step 1** 

_Let’s execute TestNG classes in parallel using TestNG xml as below:_

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE suite SYSTEM "http://testng.org/testng-1.0.dtd" >

<suite name="TestNG" parallel="methods" thread-count="2">
    <test name="Example Test" preserve-order="false">
        <classes>
            <class name="SongsTests"/>
            <class name="LoginTests"/>
        </classes>
    </test>
</suite>
```

![image](https://user-images.githubusercontent.com/104761482/221013286-485cc45d-e766-4742-a21b-dc2e10876e65.png)

**Step 2** 

_Let’s declare a static WebDriver as ThreadLocal. I have created simple methods to set, get and remove driver._

```
public class BaseTest {

  private static final ThreadLocal<WebDriver> THREAD_LOCAL = new ThreadLocal<>();
  public String url = "https://bbb.testpro.io/";
  private WebDriver driver;

  public static WebDriver getThreadLocal() {
    return THREAD_LOCAL.get();
  }

  @BeforeMethod
  public void setUpBrowser(@Optional String baseURL) throws MalformedURLException {
    THREAD_LOCAL.set(pickBrowser("firefox"));
    THREAD_LOCAL.get().manage().timeouts().implicitlyWait(Duration.ofSeconds(5));
    THREAD_LOCAL.get().manage().window().maximize();
    THREAD_LOCAL.get().manage().deleteAllCookies();
    getThreadLocal().get(url);
    System.out.println(
        "Browser setup by Thread " + Thread.currentThread().getId() + " and Driver reference is : " + getThreadLocal());

  }

  public WebDriver lambdaTest() throws MalformedURLException {
    String username = "ksenia.strigkova";
    String authkey = "mEL3Wj73VBBHOev61MB5H5WP4IiQ4xnXSBkVWO9qUrHM0RSqfk";
    String hub = "@hub.lambdatest.com/wd/hub";

    DesiredCapabilities caps = new DesiredCapabilities();
    caps.setCapability("platform", "Windows 10");
    caps.setCapability("browserName", "Chrome");
    caps.setCapability("version", "110.0");
    caps.setCapability("resolution", "1024x768");
    caps.setCapability("build", "TestNG With Java");
    caps.setCapability("name", this.getClass().getName());
    caps.setCapability("plugin", "git-testng");

    return new RemoteWebDriver(new URL("https://" + username + ":" + authkey + hub), caps);
  }

  public WebDriver pickBrowser(String browser) throws MalformedURLException {
    DesiredCapabilities capabilities = new DesiredCapabilities();
    String gridURL = "http://10.2.127.17:4444";

    switch (browser) {
      case "firefox":
        WebDriverManager.firefoxdriver().setup();
        FirefoxOptions optionsFirefox = new FirefoxOptions();
        optionsFirefox.addArguments("-private");
        return driver = new FirefoxDriver(optionsFirefox);
      case "edge":
        WebDriverManager.edgedriver().setup();
        return driver = new EdgeDriver();
      case "grid-firefox":
        capabilities.setCapability("browserName", "firefox");
        return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
      case "grid-edge":
        capabilities.setCapability("browserName", "edge");
        return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
      case "grid-chrome":
        capabilities.setCapability("browserName", "chrome");
        return driver = new RemoteWebDriver(URI.create(gridURL).toURL(), capabilities);
      case "cloud":
        return lambdaTest();
      default:
        WebDriverManager.chromedriver().setup();
        ChromeOptions optionsChrome = new ChromeOptions();
        optionsChrome.addArguments("--incognito");
        return driver = new ChromeDriver(optionsChrome);
    }
  }

  @AfterMethod
  public void tearDown() {
    THREAD_LOCAL.get().close();
    THREAD_LOCAL.remove();
  }
}
```

**Step 3** 

![image](https://user-images.githubusercontent.com/104761482/221013657-e59b8d71-ba07-48f7-bf0e-909f0afd16af.png)